torch/csrc/autograd/compiled_autograd.h

```diff
@@ -0,0 +1,583 @@
+#pragma once
+#include <c10/core/impl/TorchDispatchModeTLS.h>
+#include <torch/csrc/autograd/engine.h>
+#include <torch/csrc/utils/python_stub.h>
+#include <torch/csrc/utils/torch_dispatch_mode.h>
+#include <iostream>
+#include <typeindex>
+#include <vector>
+
+// see [Note: Compiled Autograd]
+
+namespace torch {
+namespace autograd {
+namespace generated {
+struct TypeAndSize;
+}
+
+struct SizeInput {
+  enum DynType : uint8_t { STATIC = 0, DYNAMIC = 1 };
+  SizeInput(DynType dt, int64_t v) : dyn_type(dt), value(v) {}
+  DynType dyn_type;
+  int64_t value;
+};
+
+struct CacheKeyBuffer {
+  CacheKeyBuffer() : data(nullptr) {}
+  CacheKeyBuffer(const uint8_t* key, uint16_t len) : data(new uint8_t[len]) {
+    std::memcpy(data, key, len);
+  }
+  CacheKeyBuffer(CacheKeyBuffer&& other)
+      : data(std::exchange(other.data, nullptr)) {}
+  ~CacheKeyBuffer() {
+    if (data != nullptr) {
+      delete[] data;
+    }
+  }
+  CacheKeyBuffer(const CacheKeyBuffer& other) = delete;
+  CacheKeyBuffer& operator=(const CacheKeyBuffer&) = delete;
+
+  uint8_t* data;
+};
+
+struct CacheKey {
+  // Key to find the next node in the shadow graph.  We use C++ RTTI for the
```

Comment:
I see why you used RTTI here, RTTI is annoying though lol (we occasionally have situations where morally the same C++ type have different RTTI because multiple dynamic libraries are dragging along their own copies of C++ stdlib symbols. Since we're only RTTI'ing the autograd function classes hopefully they will be consistent.)

Reply:
Yeah.  This avoided the need for a giant enum containing every autograd node type.  I agree it is annoying...